File: js/euler95.js
function euler95(){
    let minElement = 0, longest = 0;
    for (let i = 1; i < 1e6; i++){
        const seen = new Set();
        let n = i;
        // numbers in chain cannot exceed 1 million
        while (!seen.has(n) && n <= 1e6){
            seen.add(n);
            n = getDivisorSum(n);
        }
        // new longest chain found that returns to starting point
        if (n == i && seen.size > longest){
            longest = seen.size;
            minElement = Math.min(...seen);
        }
    }
    return minElement;
}

function getDivisorSum(n){
    let res = 1;
    let root = Math.sqrt(n);
    for (let i = 2; i <= root; i++){
        if (n % i == 0) res += i + (n / i);
    }
    if (root == Math.floor(root)) res -= root;
    return res;
}

console.log(euler95());

File: go/euler95.go
package main
import "fmt"

func getProperDivisorSum(n int) int {
    res := 1
    for i := 2; i * i < n; i++ {
        if n % i == 0 {
            res += i
            if i != n / i {
                res += n / i
            }
        }
    }
    return res
}

func main() {
    lim := 1000000
    res, maxLen := 0, 0
    for n := 220; n < lim; n++ {
        seen := make(map[int]bool)
        seen[n] = true
        curr := getProperDivisorSum(n)
        for curr != n {
            // element in chain exceeds limit, chain not applicable
            if curr > lim {
                break
            }
            if _, found := seen[curr]; found {
                break
            }
            seen[curr] = true
            curr = getProperDivisorSum(curr)
        }
        // new longest chain found that returns to starting point n
        if curr == n && len(seen) > maxLen {
            maxLen = len(seen)
            // start with n and update if values smaller than n are found
            res = n
            for num := range seen {
                if num < res {
                    res = num
                }
            }
        }
    }
    fmt.Println(res)
}
